sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -128,14 +150,111 @@ static void *_event_loop(void *arg)
+
+    if (type & SOCK_ASYNC_CONN_RECV) {
+        ssize_t res = sock_dtls_recv(sock, &session,
+                            _listen_buf, sizeof(_listen_buf),
+                            CONFIG_GCOAP_DTLS_HANDSHAKE_TIMEOUT_USEC);
+        if (res != -SOCK_DTLS_HANDSHAKE) {
+            DEBUG("gcoap: could not establish DTLS session: %d\n", res);
+            sock_dtls_session_destroy(sock, &session);
+            return;
+        }
+
+        /* If session has been previously stored and the state was HANDSHAKE_ONGOING
+        before, the handshake has been initiated internally and another thread
+        is waiting for the handshake.   */
+        session_state_t prev_state = NO_SESSION;
+        res = dtls_session_store(&session, SESSION_ESTABLISHED, &prev_state, false);
+        if (res) {
+            /* Send message to main thread to inform about established session */
+            if (prev_state == HANDSHAKE_ONGOING) {
+                msg_t msg = { .type = DTLS_EVENT_CONNECTED };
+                msg_send(&msg, _main_thread);
+            }
+        } else {
+            /* No space in session management. Should not happen. If it occurs,
+            we lost track of sessions stored in tinydtls internally */
+            sock_dtls_session_destroy(sock, &session);
+        }
+
+        /* if no session slot is left: set timeout to free up session */
+        if (dtls_session_get_num_available_slots() == 0)
+        {
+            uint32_t timeout = US_PER_MS * CONFIG_GCOAP_DTLS_FREE_UP_SESSION_TIMEOUT_MSEC;
+            event_callback_init(&_dtls_session_free_up_tmout_cb,
+                                _dtls_free_up_session, NULL);
+            event_timeout_init(&_dtls_session_free_up_tmout, &_queue,
+                               &_dtls_session_free_up_tmout_cb.super);
+            event_timeout_set(&_dtls_session_free_up_tmout, timeout);
+        }
+    }
+
+    if (type & SOCK_ASYNC_CONN_FIN) {
+        session = *workaround_get_finished_session();
+        if (dtls_session_remove(&session) == -1) {
+            printf("gcoap: could not remove finished DTLS session\n");
+        }
+    }
+
+    if (type & SOCK_ASYNC_MSG_RECV) {
+        ssize_t res = sock_dtls_recv(sock, &session, _listen_buf,
+                                    sizeof(_listen_buf), 0);
+        if (res <= 0) {
+            DEBUG("gcoap: DTLS receive failure: %d\n", (int)res);
+            return;
+        }
+
+        _process_coap_pdu(&_sock_udp, &session.ep,  _listen_buf, res);
+    }
+}
+
+/* Timeout function to free up a session when all session slots are occupied */
+static void _dtls_free_up_session(void *arg) {
+    (void)arg;
+    sock_dtls_session_t session;
+
+    if (dtls_session_get_num_available_slots() == 0) {
+        if(dtls_session_get_oldest_used_session(&session) != -1) {
+            /* free up session */
+            dtls_session_remove(&session);
+            sock_dtls_session_destroy(&_sock_dtls, &session);
+        }
+    }
+}
+
+#else
```

Comment:
Can we get around such large `#if` blocks? Those can easily get out of hand.

Reply:
I think with using `IS_ACTIVE()` I can really drop most (maybe even all) `#if' blocks.